sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1452,11 +1452,27 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
- * Combines two adjacent [[Limit]] operators into one, merging the
- * expressions into one single expression.
+ * 1. Eliminate [[Limit]] operators if it's child max row <= limit.
+ * 2. Combines two adjacent [[Limit]] operators into one, merging the
+ *    expressions into one single expression.
  */
-object CombineLimits extends Rule[LogicalPlan] {
-  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+object EliminateLimits extends Rule[LogicalPlan] {
+  private def canEliminate(limitExpr: Expression, child: LogicalPlan): Boolean = {
+    // We skip such case that Sort is after Limit since
+    // SparkStrategies will convert them to TakeOrderedAndProjectExec
+    val skipEliminate = child match {
+      case Sort(_, true, _) => true
```

Comment:
does `TakeOrderedAndProjectExec` really help if we need to get and sort all the output?

Reply:
It's a good question. I make a simple benchmark for this case.

```
    runBenchmark("Sort and Limit") {
      val N = 10000
      val benchmark = new Benchmark("benchmark sort and limit", N)

      benchmark.addCase("TakeOrderedAndProject", 3) { _ =>
        spark.range(N).toDF("c").sort("c").take(20000)
      }

      benchmark.addCase("Sort And Limit", 3) { _ =>
        withSQLConf("spark.sql.execution.topKSortFallbackThreshold" -> "-1") {
          spark.range(N).toDF("c").sort("c").take(20000)
        }
      }

      benchmark.addCase("Sort", 3) { _ =>
        spark.range(N).toDF("c").sort("c").collect()
      }
      benchmark.run()
    }
```

And the result is:
```
================================================================================================
Sort and Limit
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.15.6
Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
benchmark sort and limit:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
TakeOrderedAndProject                               181            210          29          0.1       18148.6       1.0X
Sort And Limit                                       90            109          16          0.1        8996.1       2.0X
Sort                                                 57             63           9          0.2        5678.5       3.2X

```

Seems `TakeOrderedAndProject` have the redundant loop in this case.

Reply:
Refer to this result, we can remove the `Limit` even if it's child is `Sort`.

Reply:
`TakeOrderedAndProjectExec` doesn't support whole stage codegen, and we expect it to be faster by avoiding sorting the entire data set. Now if must sort the entire data set, I think `TakeOrderedAndProjectExec` is actually slower. cc @viirya @maropu

Reply:
`TakeOrderedAndProjectExec` avoid range shuffle but `Limit` + `Sort` or `Sort` requires that. In the benchmark `RangeExec` already outputs in range partition. Can we insert a shuffle after `range` to break the range partition?

Reply:
Logically, you are right, but it seems something wrong with this 
`spark.range(100000).toDF("c").sort("c").explain(true)`
```
== Optimized Logical Plan ==
Sort [c#18L ASC NULLS FIRST], true
+- Project [id#16L AS c#18L]
   +- Range (0, 100000, step=1, splits=Some(2))

== Physical Plan ==
*(2) Sort [c#18L ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(c#18L ASC NULLS FIRST, 200), true, [id=#56]
   +- *(1) Project [id#16L AS c#18L]
      +- *(1) Range (0, 100000, step=1, splits=2)
```

and this case is what you said
`spark.range(100000).sort("id").explain(true)`
```
== Optimized Logical Plan ==
Range (0, 100000, step=1, splits=Some(2))

== Physical Plan ==
*(1) Range (0, 100000, step=1, splits=2)
```

Reply:
BTW, also add the shuffle benchmark
```
    runBenchmark("Sort and Limit") {
      val N = 100000
      val benchmark = new Benchmark("benchmark sort and limit", N)

      benchmark.addCase("TakeOrderedAndProject", 3) { _ =>
        spark.range(N).toDF("c").repartition(200).sort("c").take(200000)
      }

      benchmark.addCase("Sort And Limit", 3) { _ =>
        withSQLConf("spark.sql.execution.topKSortFallbackThreshold" -> "-1") {
          spark.range(N).toDF("c").repartition(200).sort("c").take(200000)
        }
      }

      benchmark.addCase("Sort", 3) { _ =>
        spark.range(N).toDF("c").repartition(200).sort("c").collect()
      }
      benchmark.run()
    }
```

and the result is 
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.15.6
Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
benchmark sort and limit:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
TakeOrderedAndProject                              1833           2259         382          0.1       18327.1       1.0X
Sort And Limit                                     1417           1658         285          0.1       14167.5       1.3X
Sort                                               1324           1484         225          0.1       13238.3       1.4X
```

Reply:
I have the same feeling with @viirya now. It seems the performance depends on shuffle overheads (data size), so   IMO its better to keep the current behaviour in this PR (so that we could control it, `TakeOrderedAndProjectExec or LIMIT+SORT`, by `topKSortFallbackThreshold`).

Reply:
I'm ok to skip eliminate this case if we afraid the potential regression. But the benchmark show it's no harmful, I also believe that, for large data size range shuffle sort is faster than singe partition sort.

And to clarify, we can choose the follow option
1. skip eliminate limit if child is sort, to avoid potential regression about `TakeOrderedAndProjectExec`.
2. eliminate all matched limit, since benchmark support it.
3. add a config to control the eliminate limit ? e.g. `spark.sql.optimizer.eliminateLimits.enable`.

Reply:
This reminds me that we can optimize root `SortExec` by collecting data to the driver and sort locally, to avoid the shuffle, if the input data size is small. @ulysses-you do you have interests to run some experiments?